mautic/mautic
# Csrf errors ajax endpoints

[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bug fix? | Y
| New feature? | N
| Automated tests included? | N
| Related user documentation PR URL | N/A
| Related developer documentation PR URL | N/A
| Issues addressed (#s or URLs) | N/a
| BC breaks? | N
| Deprecations? | N

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

[//]: # ( Required: )
#### Description:
If the `mauticAjaxCsrf` set in the user browser differs from the one set on the server, we'll get an error message whenever an AJAX action is performed. This is because we use `POST` for all AJAX requests _even if the request is idempotent_ (not expected to have side-effects).  To fix that, this PR adds support to specify the request method when calling `Mautic.ajaxActionRequest`. It also updates any idempotent AJAX requests to use `GET` instead. 

This PR does _not_ address CSRF token errors that occur when submitting a form. From our extensive research into the errors, we could not find a code bug that would cause those errors. 

[//]: # ( As applicable: )
#### Steps to reproduce the bug:
1. Login to your Mautic instance. This instance should have multiple segments or emails created in it. (We're going to reproduce using the stats column of one of those)
2. Open the developer console (right click anywhere, then click "Inspect" and go to the "Console" tab).
3. Copy and paste the following into your console `mauticAjaxCsrf = mauticAjaxCsrf + "a";`. This will change the CSRF token that gets sent with all AJAX requests.
4. Navigate to the Segments or Emails page
5. As it loads, you'll see the CSRF token error flash on the screen
6. Switch to the "Networking" tab in your developer console
7. You'll see all the stat requests were sent with `POST`

#### Steps to test this PR:
1. Redo reproduction steps
2. At step 5, no CSRF errors will appear while loading the stats column
3. At step 7, you'll see that the requests are sent using `GET`.

<a href="https://gitpod.io/#https://github.com/mautic/mautic/pull/10567"><img src="https://gitpod.io/button/open-in-gitpod.svg"/></a>
